pkg/util/pull/pull.go

```diff
@@ -15,14 +15,113 @@ import (
+// Argument pull should subscribe to changes, blocking until the subscription fails.
+// Argument get should get the current state, blocking until the state is got or an error occurs.
+// OrPoll blocks until the context is cancelled, retrying either pull or get on error.
+func OrPoll(ctx context.Context, pull, get Getter, opts ...Option) error {
 	conf := calcOpts(opts...)
 
-	shouldPoll := false
+	return runBlocking(ctx, "blocking tasks", func(ctx context.Context) error {
+		err := runPull(ctx, pull, conf)
+		if shouldReturn(err) {
+			return err
+		}
+		// the runPull err is never nil, so if we're here then we need to try polling
+		err = get(ctx) // get one to see if this is unimplemented too
+		if shouldReturn(err) {
+			return err
+		}
+		if !isUnimplemented(err) {
+			// pull is unimplemented but get is not, there's a chance things have changed (i.e. during boot)
+			// so try pull again just to be sure.
+			pullErr := runPull(ctx, pull, conf)
+			if shouldReturn(pullErr) {
+				return pullErr
+			}
+			// pull is still unimplemented, let's start the poll loop for real
+			err = runPoll(ctx, err, get, conf)
+			if shouldReturn(err) {
+				return err
+			}
+			// else get is unimplemented
+		}
+
+		// if we're here both pull and get returned unimplemented, so retry everything after a delay
+		return err
+	}, conf, func(error) bool { return false })
+}
+
+// Changes calls Pull on poller unless it's not supported, in which case it polls.
+// It will retry on error, backing off exponentially up to a maximum delay.
+// It will return if the context is cancelled or a non-recoverable error occurs.
+func Changes[C any](ctx context.Context, poller Fetcher[C], changes chan<- C, opts ...Option) error {
+	return OrPoll(ctx,
+		func(ctx context.Context) error {
+			return poller.Pull(ctx, changes)
+		},
+		func(ctx context.Context) error {
+			return poller.Poll(ctx, changes)
+		},
+		opts...)
+}
+
+// runPull retries calling pull returning on context cancellation or when pull returns an error that has codes.Unimplemented.
+// runPull will never return a nil error.
+func runPull(ctx context.Context, pull Getter, conf changeOpts) error {
+	return runBlocking(ctx, "pulls", pull, conf, isUnimplemented)
+}
+
+// runPoll blocks repeatedly calling get after some delay until context is cancelled or get returns an error that is not codes.Unimplemented.
+// runPoll never returns a nil error.
+func runPoll(ctx context.Context, err error, get Getter, conf changeOpts) error {
+	pollDelay := conf.pollDelay
+	errCount := 0
+	ticker := conf.clock.NewTicker(conf.pollDelay)
+	defer ticker.Stop()
 
+	incErr := func() {
+		errCount++
+		pollDelay = time.Duration(float64(pollDelay) * 1.2)
+		if pollDelay > conf.fallbackMaxDelay {
+			pollDelay = conf.fallbackMaxDelay
+		}
+		ticker.Reset(pollDelay)
+		if errCount == 5 {
+			conf.logger.Warn("poll is failing, will try keep retrying", zap.Error(err))
+		}
```

Comment:
Maybe use `github.com/cenkalti/backoff/v4` (which we already depend on) instead of manual backoff algorithm.

Reply:
I'll have a look, I was hoping to keep as much of the code the same as I could. Not sure what backoff's api looks like (or how it'll affect the tests)

Reply:
This has been done now.